packages/packherd-require/README.md

```diff
@@ -0,0 +1,125 @@
+function of [this type][GetModuleKey] in order to resolve those keys. 
+
+For example in the case of [v8-snapshot][v8-snapshot] (TODO: Update this link when snapshot module is added) the [getModuleKey
+implementation][v8-snapshot module key] (TODO: Update this link when snapshot module is added) implementation relies on a resolver map that is
+embedded inside the app's snapshot. Additionally it knows how modules are keyed via the
+[modified esbuild][cypress esbuild] bundler it uses.
+
+Once the module key has been resolved (or even if not) packherd tries its best to resolve
+and/or load the module from the most efficient source. It attempts to avoid accessing the file
+system until no more options remain and only loads it via the Node.js resolution/loader
+mechanism when all else failed.
+
+For more details on the module resolve/load steps refer to [PackherdModuleLoader][packherd
+module loader], in particular [`tryLoad`][try load] and [`tryResolve`][try resolve] including
+the relevant code sections which include detailed comments for each step.
+
+
+## Transpiling TypeScript Modules on Demand
+
+To enable this feature the [packherdRequire][require fn] has to be invoked in order to
+have it hook into Node.js `require` calls via a `Module._extension`. Particularly the
+[`transpileOpts`][transpile opts] field of the [opts][require opts] needs to be configured as follows.
+
+- `supportTS`: `true`
+- `initTranspileCache`: needs to be a function matching [InitTranspileCache][init transpile cache fn]
+
+### Transpile Cache
+
+We recommend to use the [dirt-simple-file-cache][dirt-simple-file-cache] module to provide the
+transpile cache as it has been developed alongside packherd for just this purpose.
+
+Here is an example of how that option field could be setup with this module.
+
+```js
+const dirtSimpleFileCache = require('dirt-simple-file-cache')
+const initTranspileCache = () => 
+  DirtSimpleFileCache.initSync(projectBaseDir, { keepInMemoryCache: true })
+```
+
+### Sourcemap Support
+
+In order to show original locations for errors logged to the console, packherd hooks into the
+generation of error stack traces and maps locations to TypeScript.
+
+For more information please read the [sourcemap docs][sourcemap docs]
+
+### Implementation
+
+Please find more implementation details regarding transpilation inside
+[./src/transpile-ts.ts][transpile-ts].
+
+### Import Caveats
+
+Since esbuild enforces the behaviour of imports being static this caused problems
+with tests that relied on being able to patch/`sinon.stub` modules even after they were
+imported. 
+
+In general we would recommend doing this _properly_ via a tool like
+[proxyquire][proxyquire].
+
+## Env Vars
+
+- `PACKHERD_CODE_FRAMES` if set will include code snippets for error messages that have been
+  sourcemapped
+
+[require fn]:https://github.com/cypress-io/cypress/blob/develop/packages/packherd/src/require.ts#L71
```

Comment:
I feel like these are going to quickly become out of date since they are referencing specific lines in develop.

Reply:
@ryanthemanuel FYI you can get a permalink to a file (based off the commit sha and not the branch name) in GitHub by typing `Y` when looking at a file. This might need to be done as a followup after this is merged. That would lead to a risk the linked code will be out-of-date with develop, but at least it won't potentially be pointing to something complete different.

Reply:
Yeah that was my main thought. I used to have permalinks but they were from (soon to be deleted) branches. I think I'll do a quick follow up PR with the develop sha afterwards.